File: frontend/lib/data/decision-review.ts
/**
 * Your own decisions, scored against what the engine said before the deadline.
 *
 * ## Why this is not another accuracy screen
 *
 * `accuracy.json` asks whether the model was right. This asks whether *you* were,
 * and it can only ask honestly because `predictions/fpl/ledger/gwNN/forecast.jsonl`
 * records a forecast with a `sealed_at` and a `seconds_before_deadline`. Every
 * verdict below is therefore a claim about what was knowable at the time rather
 * than hindsight wearing a confident face.
 *
 * ## The three verdicts, and why a tie is one of them
 *
 * A bench call is `foreseeable`, `defensible` or `indistinguishable`. The third is
 * the one that matters: the producer compares two sealed projections against the
 * Monte Carlo standard error of each, so two players the simulation could not
 * separate never become a lesson. Rendering `indistinguishable` as a mild version
 * of `foreseeable` would manufacture a weekly reproach out of arithmetic noise —
 * on GW1 the one real bench cost was a 0.0008 xP gap against a combined 2-sigma of
 * 0.100, and a naive `>` would have called it an error.
 *
 * A missing verdict (`null`) is a fourth state and not a fourth judgement: the
 * sealed universe excludes fringe players by design, so a player it never covered
 * has no forecast to be judged against. Absence of a verdict is absence of
 * evidence, exactly as in `minutes-conflicts.ts`.
 *
 * ## What it withholds
 *
 * `aggregate` is null until `minimumObservations` gameweeks have settled, and
 * `aggregateReason` says so in the producer's own words. One week of
 * self-assessment is noise, and a rate rendered off it is indistinguishable from a
 * real finding. Per-gameweek calls always ship, because "three of your four bench
 * players were rated above the man you started" is a fact about one gameweek and
 * needs no season behind it.
 */

import { malformed, narrowed, type NarrowResult } from "@/lib/data/artifact";
import {
  optBoolean,
  optNumber,
  optString,
  Problems,
  reqArray,
  reqRecord,
} from "@/lib/data/check";
import { DAY, type Descriptor } from "@/lib/data/registry";

/** How a bench or selection call stands against the sealed forecast. */
export type Verdict = "foreseeable" | "defensible" | "indistinguishable";

/** What became of a benched player. */
export type BenchKind = "rescued" | "cost" | "correct" | "no_claim";

const VERDICTS: readonly Verdict[] = [
  "foreseeable",
  "defensible",
  "indistinguishable",
];
const KINDS: readonly BenchKind[] = ["rescued", "cost", "correct", "no_claim"];

export interface BenchCall {
  readonly benchElement: number;
  /** Null for `rescued` without a pair, and for `correct` / `no_claim`. */
  readonly starterElement: number | null;
  readonly kind: BenchKind;
  /** Points the benched player beat the starter by. Zero is a real zero here. */
  readonly pointsForgone: number;
  /** Null when the sealed forecast did not cover both players. */
  readonly verdict: Verdict | null;
  /**
   * The producer's own judgement, carried rather than recomputed.
   *
   * Deliberately not derived here as `verdict === "foreseeable"`: the rule is a
   * producer decision (a rescued error still counts, because the call was the same
   * whether or not an automatic substitution covered it) and a second
   * implementation on this side would be free to drift from it.
   */
  readonly isLesson: boolean;
}

export interface Selection {
  readonly worstStarter: number | null;
  readonly bestBench: number | null;
  /** Benched players the forecast separated above the weakest starter. */
  readonly benchRatedHigher: readonly number[];
  readonly gap: number | null;
  readonly misordered: boolean;
}

export interface CaptainCall {
  readonly chosen: number;
  readonly sealedBest: number | null;
  readonly agreed: boolean | null;
  /** Doubled, because the armband doubles. Positive means the engine won. */
  readonly pointsDelta: number | null;
}

export interface GameweekReview {
  readonly gameweek: number;
  /** Proof this was a pre-deadline forecast. Null means the seal said nothing. */
  readonly sealedAt: string | null;
  readonly secondsBeforeDeadline: number | null;
  readonly sealedUniverse: number | null;
  readonly points: number | null;
  /**
   * FPL's own bench figure, which is computed AFTER automatic substitutions and
   * therefore cannot see a rescue. Kept beside ours precisely because they differ.
   */
  readonly fplPointsOnBench: number | null;
  readonly transfers: number | null;
  readonly hitCost: number | null;
  readonly submittedEleven: readonly number[];
  readonly submittedBench: readonly number[];
  readonly selection: Selection | null;
  readonly bench: readonly BenchCall[];
  readonly captain: CaptainCall | null;
}

export interface Aggregate {
  readonly gameweeks: number;
  readonly pointsForgoneOnBench: number;
  readonly foreseeableBenchErrors: number;
  readonly captainAgreementRate: number | null;
  readonly captainPointsVsEngine: number | null;
}

export interface DecisionReview {
  readonly generatedAt: string | null;
  /**
   * The producer's schema version.
   *
   * Narrowed rather than left out so the provenance strip names the writer instead
   * of printing "version unknown" beside a real judgement about your season — the
   * exact drift `health.json` demonstrated, where a complete, fresh file from a
   * producer that emitted no version looked identical to a current one.
   */
  readonly producerVersion: string | null;
  readonly season: string | null;
  readonly teamName: string | null;
  readonly observations: number;
  readonly minimumObservations: number;
  /** Null until the sample is large enough. Never a partial number. */
  readonly aggregate: Aggregate | null;
  /** The producer's stated reason for withholding. Rendered verbatim. */
  readonly aggregateReason: string | null;
  readonly gameweeks: readonly GameweekReview[];
  /** Element id (as a string, because JSON keys are) to display name. */
  readonly names: ReadonlyMap<string, string>;
}

function narrowVerdict(raw: unknown): Verdict | null {
  const value = optString(raw);
  return value !== null && VERDICTS.includes(value as Verdict)
    ? (value as Verdict)
    : null;
}

function narrowBench(raw: unknown): BenchCall | null {
  const problems = new Problems();
  const row = reqRecord(raw, "bench call", problems);
  if (!row) return null;

  const benchElement = optNumber(row.bench_element);
  const kind = optString(row.kind);
  // Without an element there is nobody to name, and without a kind there is no
  // claim being made. Either missing makes the row unrenderable rather than
  // partial.
  if (benchElement === null || kind === null) return null;
  if (!KINDS.includes(kind as BenchKind)) return null;

  return {
    benchElement,
    starterElement: optNumber(row.starter_element),
    kind: kind as BenchKind,
    pointsForgone: optNumber(row.points_forgone) ?? 0,
    verdict: narrowVerdict(row.verdict),
    isLesson: optBoolean(row.is_lesson) ?? false,
  };
}

function narrowSelection(raw: unknown): Selection | null {
  if (raw === null || raw === undefined) return null;
  const problems = new Problems();
  const row = reqRecord(raw, "selection", problems);
  if (!row) return null;
  const higher = Array.isArray(row.bench_rated_higher)
    ? row.bench_rated_higher.filter((n): n is number => typeof n === "number")
    : [];
  return {
    worstStarter: optNumber(row.worst_starter),
    bestBench: optNumber(row.best_bench),
    benchRatedHigher: higher,
    gap: optNumber(row.gap),
    // Derived from the list rather than trusted separately, so a payload whose
    // flag and list disagree cannot render a contradiction.
    misordered: higher.length > 0,
  };
}

function narrowCaptain(raw: unknown): CaptainCall | null {
  if (raw === null || raw === undefined) return null;
  const problems = new Problems();
  const row = reqRecord(raw, "captain", problems);
  if (!row) return null;
  const chosen = optNumber(row.chosen);
  if (chosen === null) return null;
  return {
    chosen,
    sealedBest: optNumber(row.sealed_best),
    agreed: optBoolean(row.agreed),
    pointsDelta: optNumber(row.points_delta),
  };
}

function narrowGameweek(raw: unknown): GameweekReview | null {
  const problems = new Problems();
  const row = reqRecord(raw, "gameweek review", problems);
  if (!row) return null;
  const gameweek = optNumber(row.gameweek);
  if (gameweek === null) return null;

  const ids = (value: unknown): readonly number[] =>
    Array.isArray(value)
      ? value.filter((n): n is number => typeof n === "number")
      : [];

  return {
    gameweek,
    sealedAt: optString(row.sealed_at),
    secondsBeforeDeadline: optNumber(row.seconds_before_deadline),
    sealedUniverse: optNumber(row.sealed_universe),
    points: optNumber(row.points),
    fplPointsOnBench: optNumber(row.fpl_points_on_bench),
    transfers: optNumber(row.transfers),
    hitCost: optNumber(row.hit_cost),
    submittedEleven: ids(row.submitted_eleven),
    submittedBench: ids(row.submitted_bench),
    selection: narrowSelection(row.selection),
    bench: (Array.isArray(row.bench) ? row.bench : [])
      .map(narrowBench)
      .filter((c): c is BenchCall => c !== null),
    captain: narrowCaptain(row.captain),
  };
}

function narrowAggregate(raw: unknown): Aggregate | null {
  if (raw === null || raw === undefined) return null;
  const problems = new Problems();
  const row = reqRecord(raw, "aggregate", problems);
  if (!row) return null;
  const gameweeks = optNumber(row.gameweeks);
  if (gameweeks === null) return null;
  return {
    gameweeks,
    pointsForgoneOnBench: optNumber(row.points_forgone_on_bench) ?? 0,
    foreseeableBenchErrors: optNumber(row.foreseeable_bench_errors) ?? 0,
    captainAgreementRate: optNumber(row.captain_agreement_rate),
    captainPointsVsEngine: optNumber(row.captain_points_vs_engine),
  };
}

export function narrowDecisionReview(raw: unknown): NarrowResult<DecisionReview> {
  const problems = new Problems();
  const file = reqRecord(raw, "decision_review", problems);
  if (!file) return malformed(problems.all);

  const rows = reqArray(file.gameweeks, "gameweeks", problems);
  if (!rows) return malformed(problems.all);

  const observations = optNumber(file.observations);
  const minimum = optNumber(file.minimum_observations);
  // Both are required, not defaulted. A withheld aggregate is only interpretable
  // against the bar it failed to clear, and inventing a threshold here would let
  // the producer change its mind without the page noticing — the same reason
  // `minutes-conflicts.ts` refuses to default its own thresholds.
  if (observations === null || minimum === null) return malformed(problems.all);

  const names = new Map<string, string>();
  if (file.names && typeof file.names === "object") {
    for (const [key, value] of Object.entries(file.names)) {
      if (typeof value === "string") names.set(key, value);
    }
  }

  const producerVersion = optNumber(file.producer_version);

  return narrowed({
    generatedAt: optString(file.generated_at),
    producerVersion:
      producerVersion === null ? null : String(producerVersion),
    season: optString(file.season),
    teamName: optString(file.team_name),
    observations,
    minimumObservations: minimum,
    aggregate: narrowAggregate(file.aggregate),
    aggregateReason: optString(file.aggregate_reason),
    gameweeks: rows
      .map(narrowGameweek)
      .filter((g): g is GameweekReview => g !== null),
    names,
  });
}

/**
 * Nothing has settled yet.
 *
 * A real and expected state for most of August, and declared rather than inferred
 * so the page can say "no gameweek has settled" instead of rendering an absence.
 */
export function decisionReviewIsEmpty(value: DecisionReview): boolean {
  return value.gameweeks.length === 0;
}

/** A player's display name, or null. Never the raw id dressed as a name. */
export function nameOf(
  review: DecisionReview,
  element: number | null,
): string | null {
  if (element === null) return null;
  return review.names.get(String(element)) ?? null;
}

/** Every call the forecast says was avoidable, newest gameweek first. */
export function lessons(
  review: DecisionReview,
): readonly { readonly gameweek: number; readonly call: BenchCall }[] {
  return [...review.gameweeks]
    .sort((a, b) => b.gameweek - a.gameweek)
    .flatMap((week) =>
      week.bench
        .filter((call) => call.isLesson)
        .map((call) => ({ gameweek: week.gameweek, call })),
    );
}

export const DECISION_REVIEW: Descriptor<DecisionReview> = {
  key: "decisionReview",
  path: "fpl/decision_review.json",
  owner: "agent",
  describes: "your own decisions, scored against the pre-deadline seal",
  // Rewritten when a gameweek settles, which is weekly at most. A day-old copy is
  // entirely normal here, so the budget is generous — unlike the news tick, a
  // stale review is not evidence that anything has stopped.
  freshnessBudgetMs: 30 * DAY,
  narrow: narrowDecisionReview,
  producedAtOf: (v) => v.generatedAt,
  producerVersionOf: (v) => v.producerVersion,
  isEmpty: decisionReviewIsEmpty,
};

File: pipeline/learning/decision_review.py
"""
Were *you* right, and was a call wrong or merely unlucky.

Everything else in this package scores the model. This scores the manager, and it
is only honest because of one asset nothing else in the category has: a forecast
sealed before the deadline. `predictions/fpl/ledger/gw{NN}/forecast.jsonl` records
what the engine said with a `sealed_at` and a `seconds_before_deadline`, so the
counterfactual "the engine preferred your bench player" is a claim about what was
knowable at the time rather than hindsight dressed up as insight.

Take that away and this module is a scoreboard. Every other FPL product can tell
you that you left points on the bench. None can tell you which of those you could
have known, because none of them wrote a number down first.

## Three things the naive version gets wrong

**1. FPL rewrites your team before you can read it back.** The
`entry/{id}/event/{gw}/picks/` endpoint returns positions and multipliers *after*
automatic substitutions are applied. Measured on GW1 2026-27: the response showed
Thomas in an XI slot and Palestra on the bench, while `automatic_subs` recorded
Thomas coming *in* for Palestra — so the submitted eleven was the opposite of what
the payload displayed. Scoring the payload as-is grades FPL's correction of your
team instead of your team, which flatters every decision the auto-sub rescued.
`submitted_eleven` reverses them.

**2. Only the substitute who actually came on was rescued.** A first pass at this
credited every bench player who played against the same failed starter, reporting
two rescues for one substitution. The rescue set is exactly `element_in` from
`automatic_subs` and nothing else; the remaining bench players are ordinary
comparisons.

**3. A tie is not a lesson.** Two players the simulation cannot separate must never
produce a "you should have known". The sealed record publishes `mc_se`, the Monte
Carlo standard error of each mean, so the threshold is derived from the run's own
precision rather than picked: a gap counts only if it clears `sigmas` times the
combined standard error. On GW1 that correctly separated Thomas over Palestra
(gap 0.074, combined 2-sigma 0.053) while refusing a 0.0008 gap between two
midfielders whose combined 2-sigma was 0.100 — a gap `>` would have called a
lesson every week for the rest of the season.

## What it refuses to say

One settled gameweek is noise. `build` reports `observations` and withholds every
aggregate verdict below `MINIMUM_OBSERVATIONS`, the same refusal `accuracy.py`
makes for the model: a self-assessment that renders a confident number off one
week is indistinguishable from a real finding, and worse than saying nothing.

Per-gameweek calls are always emitted, because "Palestra did not play and three of
your four bench players were rated above him" is a fact about one gameweek and
does not need a season behind it. It is the *aggregate* — your captaincy edge, your
foreseeable-error rate — that needs the sample.
"""
from __future__ import annotations

import logging
import math
from dataclasses import dataclass
from typing import Any, Dict, Iterable, List, Mapping, Optional, Sequence, Set, Tuple

logger = logging.getLogger(__name__)

SCHEMA_VERSION = 1

#: Settled gameweeks required before any aggregate is reported. Below this the
#: per-gameweek calls stand alone and every rate is withheld.
MINIMUM_OBSERVATIONS = 6

#: How many combined standard errors two projections must differ by before the
#: difference is treated as something the simulation actually resolved. Two is
#: roughly a 95% interval on the difference of two means.
DEFAULT_SIGMAS = 2.0

#: A bench call's kind.
RESCUED = "rescued"          # auto-subbed in for a starter who did not play
COST = "cost"                # played, outscored a starter who also played
CORRECT = "correct"          # played, beat nobody who played
NO_CLAIM = "no_claim"        # did not play; benching him cannot be judged

#: Whether the sealed forecast supports calling a decision an error.
FORESEEABLE = "foreseeable"        # the engine ranked the bench player higher
DEFENSIBLE = "defensible"          # the engine ranked the starter higher
INDISTINGUISHABLE = "indistinguishable"  # the engine did not separate them


@dataclass(frozen=True)
class Sealed:
    """One player's pre-deadline forecast, with the precision behind it."""

    element_id: int
    xp: float
    #: Monte Carlo standard error of ``xp``. Without it a tie is unrecognisable.
    mc_se: float


@dataclass(frozen=True)
class BenchCall:
    """One bench decision, and whether the sealed forecast condemns it."""

    bench_element: int
    #: The starter this is measured against, or None for RESCUED/NO_CLAIM shapes
    #: where there is no comparison to make.
    starter_element: Optional[int]
    kind: str
    #: Points the bench player scored minus the starter's. Zero for NO_CLAIM.
    points_forgone: int
    #: One of FORESEEABLE / DEFENSIBLE / INDISTINGUISHABLE, or None when the
    #: sealed forecast did not cover both players — absent, never assumed.
    verdict: Optional[str]

    def is_lesson(self) -> bool:
        """
        The forecast said this was avoidable, whether or not it cost anything.

        Deliberately not gated on ``points_forgone``. Starting a player the engine
        ranked below one you benched is the same decision whether an automatic
        substitution rescued it or not — GW1 2026-27 produced exactly that case, and
        counting only the expensive ones would teach the manager to notice bad calls
        solely when they were also unlucky. The cost is reported separately.
        """
        return self.verdict == FORESEEABLE


@dataclass(frozen=True)
class CaptainCall:
    """Your armband against the sealed argmax of the eleven you submitted."""

    chosen: int
    #: Highest sealed xp among the submitted eleven, or None when the forecast
    #: covered none of them.
    sealed_best: Optional[int]
    agreed: Optional[bool]
    #: Actual points the argmax scored minus what your captain scored, doubled
    #: because the armband doubles. Positive means the engine's pick was better.
    #: None when either side is unknown.
    points_delta: Optional[int]


def separated(
    left: Sealed,
    right: Sealed,
    *,
    sigmas: float = DEFAULT_SIGMAS,
) -> bool:
    """
    Did the simulation actually resolve a difference between these two?

    The standard error of a difference of independent means is the root sum of
    squares of theirs. Anything inside ``sigmas`` of that is noise the run cannot
    distinguish, and calling it a managerial error would manufacture a lesson from
    arithmetic.
    """
    gap = abs(left.xp - right.xp)
    combined = math.sqrt(left.mc_se**2 + right.mc_se**2)
    return gap > sigmas * combined


def _verdict(
    bench: Optional[Sealed],
    starter: Optional[Sealed],
    *,
    sigmas: float = DEFAULT_SIGMAS,
) -> Optional[str]:
    """
    Whether the forecast condemns preferring ``starter`` over ``bench``.

    None when either player is missing from the sealed universe — its criteria
    exclude fringe players by design, and an absent forecast is not a forecast of
    zero.
    """
    if bench is None or starter is None:
        return None
    if not separated(bench, starter, sigmas=sigmas):
        return INDISTINGUISHABLE
    return FORESEEABLE if bench.xp > starter.xp else DEFENSIBLE


def submitted_eleven(picks: Mapping[str, Any]) -> Tuple[Set[int], List[int]]:
    """
    The eleven the manager actually submitted, and the four they benched.

    ``picks`` is the raw ``entry/{id}/event/{gw}/picks/`` payload, whose
    ``multiplier`` already reflects automatic substitutions. Reversing them is what
    recovers the decision: each substitution took ``element_out`` out of the
    submitted eleven and put ``element_in`` in, so undoing it restores both.

    Returns the eleven as a set and the bench in the payload's own position order,
    because bench order is itself a decision and the caller may want to score it.
    """
    rows = picks.get("picks") or []
    eleven = {
        int(row["element"])
        for row in rows
        if isinstance(row, Mapping) and int(row.get("multiplier") or 0) > 0
    }
    for sub in picks.get("automatic_subs") or []:
        if not isinstance(sub, Mapping):
            continue
        came_on, went_off = sub.get("element_in"), sub.get("element_out")
        if came_on is None or went_off is None:
            # A half-recorded substitution cannot be reversed. Leaving the eleven
            # as FPL rewrote it is the safer failure: it under-reports the
            # manager's errors rather than inventing one.
            logger.warning("automatic_sub missing element_in/out; not reversed: %s", sub)
            continue
        eleven.discard(int(came_on))
        eleven.add(int(went_off))

    bench = [
        int(row["element"])
        for row in sorted(
            (r for r in rows if isinstance(r, Mapping)),
            key=lambda r: int(r.get("position") or 0),
        )
        if int(row["element"]) not in eleven
    ]
    return eleven, bench


def rescued_pairs(picks: Mapping[str, Any]) -> Dict[int, int]:
    """
    Bench player -> the starter they were automatically substituted for.

    Exactly the recorded substitutions. Deriving this by looking for any failed
    starter is the mistake documented at the top of this module.
    """
    pairs: Dict[int, int] = {}
    for sub in picks.get("automatic_subs") or []:
        if not isinstance(sub, Mapping):
            continue
        came_on, went_off = sub.get("element_in"), sub.get("element_out")
        if came_on is not None and went_off is not None:
            pairs[int(came_on)] = int(went_off)
    return pairs


def review_bench(
    picks: Mapping[str, Any],
    points: Mapping[int, int],
    minutes: Mapping[int, int],
    sealed: Mapping[int, Sealed],
    *,
    sigmas: float = DEFAULT_SIGMAS,
) -> List[BenchCall]:
    """
    Every bench decision, classified.

    ``points`` and ``minutes`` are realised per element, from ``event/{gw}/live/``.
    A bench player who did not play yields NO_CLAIM rather than a zero cost: he
    might have been the right man to bench or the wrong one, and the gameweek did
    not say.
    """
    eleven, bench = submitted_eleven(picks)
    rescues = rescued_pairs(picks)
    played_starters = [e for e in eleven if minutes.get(e, 0) > 0]

    calls: List[BenchCall] = []
    for element in bench:
        if element in rescues:
            starter = rescues[element]
            calls.append(
                BenchCall(
                    bench_element=element,
                    starter_element=starter,
                    kind=RESCUED,
                    # The substitution already collected these points, so nothing
                    # was forgone. The verdict is still worth stating: it says
                    # whether the starter should have been picked at all.
                    points_forgone=0,
                    verdict=_verdict(
                        sealed.get(element), sealed.get(starter), sigmas=sigmas
                    ),
                )
            )
            continue

        if minutes.get(element, 0) <= 0:
            calls.append(
                BenchCall(
                    bench_element=element,
                    starter_element=None,
                    kind=NO_CLAIM,
                    points_forgone=0,
                    verdict=None,
                )
            )
            continue

        scored = points.get(element, 0)
        beaten = [s for s in played_starters if points.get(s, 0) < scored]
        if not beaten:
            calls.append(
                BenchCall(
                    bench_element=element,
                    starter_element=None,
                    kind=CORRECT,
                    points_forgone=0,
                    verdict=None,
                )
            )
            continue

        # The worst starter who actually played is the swap the manager could have
        # made, so it is the one the cost is measured against. Ties in points break
        # on the lower sealed projection, which is the more conservative charge.
        worst = min(
            beaten,
            key=lambda s: (points.get(s, 0), sealed[s].xp if s in sealed else 0.0),
        )
        calls.append(
            BenchCall(
                bench_element=element,
                starter_element=worst,
                kind=COST,
                points_forgone=scored - points.get(worst, 0),
                verdict=_verdict(sealed.get(element), sealed.get(worst), sigmas=sigmas),
            )
        )
    return calls


@dataclass(frozen=True)
class ElevenCheck:
    """
    The selection question, asked independently of who was auto-subbed.

    A per-player bench call compares a benched man against the starter he replaced
    or outscored, and both of those are decided by events. This asks the thing that
    was decided *before* kickoff: did the manager start anyone the engine ranked
    below a player they left out? On GW1 2026-27 the bench calls charged one
    marginal error against a single substitute, while this shows three of four bench
    players were rated above the weakest starter — the same decision, correctly
    sized.
    """

    worst_starter: Optional[int]
    best_bench: Optional[int]
    #: Bench players the forecast separated above the weakest starter.
    bench_rated_higher: Tuple[int, ...]
    #: Sealed xp gap between the best bench player and the worst starter, when both
    #: are covered and separated. None otherwise.
    gap: Optional[float]

    @property
    def misordered(self) -> bool:
        return bool(self.bench_rated_higher)


def review_eleven(
    picks: Mapping[str, Any],
    sealed: Mapping[int, Sealed],
    *,
    sigmas: float = DEFAULT_SIGMAS,
) -> ElevenCheck:
    """Was the submitted eleven consistent with the sealed forecast's own ordering?"""
    eleven, bench = submitted_eleven(picks)
    covered_xi = [e for e in eleven if e in sealed]
    covered_bench = [e for e in bench if e in sealed]
    if not covered_xi or not covered_bench:
        return ElevenCheck(None, None, (), None)

    worst = min(covered_xi, key=lambda e: sealed[e].xp)
    best = max(covered_bench, key=lambda e: sealed[e].xp)
    higher = tuple(
        e
        for e in covered_bench
        if sealed[e].xp > sealed[worst].xp
        and separated(sealed[e], sealed[worst], sigmas=sigmas)
    )
    gap = (
        sealed[best].xp - sealed[worst].xp
        if separated(sealed[best], sealed[worst], sigmas=sigmas)
        else None
    )
    return ElevenCheck(
        worst_starter=worst, best_bench=best, bench_rated_higher=higher, gap=gap
    )


def review_captain(
    picks: Mapping[str, Any],
    points: Mapping[int, int],
    sealed: Mapping[int, Sealed],
) -> Optional[CaptainCall]:
    """
    The armband against the engine's own pick, restricted to the submitted eleven.

    Restricted deliberately: comparing against the best captain in the whole game
    would score squad building, not the armband. The decision being reviewed is the
    one the manager actually had in front of them.
    """
    rows = picks.get("picks") or []
    chosen = next(
        (
            int(row["element"])
            for row in rows
            if isinstance(row, Mapping) and row.get("is_captain")
        ),
        None,
    )
    if chosen is None:
        return None

    eleven, _ = submitted_eleven(picks)
    covered = [e for e in eleven if e in sealed]
    if not covered:
        return CaptainCall(chosen=chosen, sealed_best=None, agreed=None, points_delta=None)

    best = max(covered, key=lambda e: sealed[e].xp)
    delta: Optional[int] = None
    if chosen in points or best in points:
        # The armband doubles, so a different captain moves the total by twice the
        # difference in their scores.
        delta = 2 * (points.get(best, 0) - points.get(chosen, 0))
    return CaptainCall(
        chosen=chosen, sealed_best=best, agreed=chosen == best, points_delta=delta
    )


def load_sealed(lines: Iterable[str]) -> Dict[int, Sealed]:
    """
    The sealed forecast, keyed by element.

    ``forecast.jsonl``'s first line is a ``record: "header"`` carrying the seal
    metadata; the rest are ``record: "forecast"`` player rows. Anything else is
    skipped rather than guessed at.
    """
    import json

    out: Dict[int, Sealed] = {}
    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            row = json.loads(line)
        except ValueError:
            logger.warning("unparseable line in sealed forecast; skipped")
            continue
        if not isinstance(row, Mapping) or row.get("record") != "forecast":
            continue
        element = row.get("element_id")
        xp = row.get("xp")
        mc_se = row.get("mc_se")
        if not isinstance(element, int) or xp is None or mc_se is None:
            # Without mc_se a tie cannot be recognised, and a default would make
            # every close call look decisive. Drop the row instead.
            continue
        out[element] = Sealed(element_id=element, xp=float(xp), mc_se=float(mc_se))
    return out


def sealed_header(lines: Iterable[str]) -> Optional[Dict[str, Any]]:
    """The seal metadata, which is what makes any of this a pre-deadline claim."""
    import json

    for line in lines:
        line = line.strip()
        if not line:
            continue
        try:
            row = json.loads(line)
        except ValueError:
            continue
        if isinstance(row, Mapping) and row.get("record") == "header":
            return dict(row)
    return None


def _aggregate(
    reviews: Sequence[Mapping[str, Any]], minimum: int
) -> Optional[Dict[str, Any]]:
    """
    Season-to-date rates, or None while the sample is too small to mean anything.

    Returning None rather than a number with a caveat is deliberate: a caveat next
    to a figure is read as the figure.
    """
    if len(reviews) < minimum:
        return None

    lessons = 0
    forgone = 0
    captain_agreements = 0
    captain_measured = 0
    captain_delta = 0
    for review in reviews:
        for call in review.get("bench") or []:
            if call.get("kind") == COST:
                forgone += int(call.get("points_forgone") or 0)
                if call.get("verdict") == FORESEEABLE:
                    lessons += 1
        captain = review.get("captain")
        if captain and captain.get("agreed") is not None:
            captain_measured += 1
            captain_agreements += 1 if captain["agreed"] else 0
            captain_delta += int(captain.get("points_delta") or 0)

    return {
        "gameweeks": len(reviews),
        "points_forgone_on_bench": forgone,
        "foreseeable_bench_errors": lessons,
        "captain_agreement_rate": (
            round(captain_agreements / captain_measured, 4) if captain_measured else None
        ),
        "captain_points_vs_engine": captain_delta if captain_measured else None,
    }


def build(
    reviews: Sequence[Mapping[str, Any]],
    *,
    generated_at: str,
    season: Optional[str] = None,
    minimum_observations: int = MINIMUM_OBSERVATIONS,
) -> Dict[str, Any]:
    """
    The published view: every gameweek's calls, and aggregates once earned.

    ``reviews`` is one entry per settled gameweek, each as produced by
    ``review_gameweek``.
    """
    aggregate = _aggregate(reviews, minimum_observations)
    return {
        "schema_version": SCHEMA_VERSION,
        # Carried so the provenance strip can name the producer rather than
        # printing "version unknown" beside a real judgement about your season.
        "producer_version": SCHEMA_VERSION,
        "generated_at": generated_at,
        "season": season,
        "observations": len(reviews),
        "minimum_observations": minimum_observations,
        "aggregate": aggregate,
        "aggregate_reason": (
            None
            if aggregate is not None
            else (
                f"{len(reviews)} settled gameweek(s) reviewed; "
                f"{minimum_observations} are needed before a rate is reported. "
                "Per-gameweek calls below are complete."
            )
        ),
        "gameweeks": list(reviews),
    }


def review_gameweek(
    gameweek: int,
    picks: Mapping[str, Any],
    points: Mapping[int, int],
    minutes: Mapping[int, int],
    sealed: Mapping[int, Sealed],
    *,
    seal: Optional[Mapping[str, Any]] = None,
    sigmas: float = DEFAULT_SIGMAS,
) -> Dict[str, Any]:
    """One gameweek's decisions, as a serialisable record."""
    bench = review_bench(picks, points, minutes, sealed, sigmas=sigmas)
    captain = review_captain(picks, points, sealed)
    selection = review_eleven(picks, sealed, sigmas=sigmas)
    history = picks.get("entry_history") or {}
    eleven, bench_order = submitted_eleven(picks)

    return {
        "gameweek": gameweek,
        # Carried so a reader can tell a pre-deadline claim from a post-hoc one
        # without opening the ledger. Absent seal metadata is stated, not implied.
        "sealed_at": (seal or {}).get("sealed_at"),
        "seconds_before_deadline": (seal or {}).get("seconds_before_deadline"),
        "sealed_universe": len(sealed) or None,
        "points": history.get("points"),
        # FPL's own figure, kept beside ours precisely because they differ: theirs
        # is computed after automatic substitutions and so cannot see a rescue.
        "fpl_points_on_bench": history.get("points_on_bench"),
        "transfers": history.get("event_transfers"),
        "hit_cost": history.get("event_transfers_cost"),
        "submitted_eleven": sorted(eleven),
        "submitted_bench": bench_order,
        "selection": {
            "worst_starter": selection.worst_starter,
            "best_bench": selection.best_bench,
            "bench_rated_higher": list(selection.bench_rated_higher),
            "gap": None if selection.gap is None else round(selection.gap, 4),
            "misordered": selection.misordered,
        },
        "bench": [
            {
                "bench_element": call.bench_element,
                "starter_element": call.starter_element,
                "kind": call.kind,
                "points_forgone": call.points_forgone,
                "verdict": call.verdict,
                "is_lesson": call.is_lesson(),
            }
            for call in bench
        ],
        "captain": (
            None
            if captain is None
            else {
                "chosen": captain.chosen,
                "sealed_best": captain.sealed_best,
                "agreed": captain.agreed,
                "points_delta": captain.points_delta,
            }
        ),
    }
